include/zephyr/sys/predef_heap.h

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * Author: Adrian Warecki <adrian.warecki@intel.com>
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_PREDEF_HEAP_H_
+#define ZEPHYR_INCLUDE_SYS_PREDEF_HEAP_H_
+
+#include <zephyr/types.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/sys/util.h>
+
+/* Simple heap implementation with predefined buffers size.
+ *
+ * The blob of memory passed to the heap will be divided into a predefined number of buffers
+ * of specified sizes. Buffers of various sizes are supported. Configuration of a memory division
+ * for a buffers can be changed in runtime. The allocator tries to allocate the smallest buffer that
+ * satisfies the request. Adjacent buffers are not merged, which eliminates the problem of memory
+ * fragmentation. Allocate and free by user pointer and not an opaque block handle. Current
+ * configuration and bookkeeping is stored in a separate memory.
+ */
+
+/* Configuration of a single bundle of buffers. Defines its count and size. */
+struct predef_heap_config {
+	size_t size;
+	int count;
+};
+
+/* predef_heap object structure */
+struct predef_heap {
+	int bundles_count;
+	void *start;
+	size_t size;
+	/* Pointer and size to memory used for configuration store and bookkeeping. */
+	void *config;
+	size_t config_size;
+};
```

Comment:
My question is - how are those blocks of memory different than [memory slabs](https://docs.zephyrproject.org/1.14.0/reference/kernel/memory/slabs.html)?
Can't we use them here for underlying implementation? It would be good to re-use existing Zephyr abstractions...

Reply:
Andy had asked that before. Let's discuss this topic in one place.